docs/Contributing/Adding-new-endpoints.md

```diff
@@ -0,0 +1,255 @@
+# API Endpoints HOWTO
+
+## Steps to add a new endpoint
+
+There are two main ways to add a new endpoint to the Fleet API:
+
+1. Building the data layer first (datastore) and then going up the stack until the API endpoint.
+2. Building the API endpoint and then going down the stack until the datastore.
+
+For the sake of ease of writing this document, we'll cover case 1. If you prefer to build in more of an option 2 style
+you can simply read the documentation bottom up.
+
+### Step 1: Datastore
+
+So let's say you want to add an endpoint that would count the total number of hosts enrolled in Fleet. A SQL query for
+gathering this data could be the following:
+
+```sql
+SELECT COUNT(*) FROM hosts
+```
+
+So let's create a function for this within the mysql datastore in the [hosts.go file](https://github.com/fleetdm/fleet/blob/main/server/datastore/mysql/hosts.go):
+
+```go
+func (ds *Datastore) CountAllHost(ctx context.Context) (int, error) {
+    var hostCount int
+    err := sqlx.GetContext(ctx, db, &hostCount, `SELECT COUNT(*) FROM hosts`)
+    if err != nil {
+        return 0, err
+    }
+    return hostCount, nil
+}
+```
+
+Now this is part of the `Datastore` struct. In order to use it in an endpoint, we need this to be exposed by the `Datastore`
+interface. So we add this method [to it](https://github.com/fleetdm/fleet/blob/main/server/fleet/datastore.go#L25):
+
+```go
+type Datastore interface {
+	// rest of the interface here
+	
+    CountAllHost(ctx context.Context) (int, error)
+}
+```
+
+Now we are ready to create a method in the service.
+
+### Step 2: Service
+
+In order to use this new Datastore function we created, the layer that is in communication with it is the `Service` 
+which is both [an interface](https://github.com/fleetdm/fleet/blob/main/server/fleet/service.go#L41) and 
+[a struct](https://github.com/fleetdm/fleet/blob/main/server/service/service.go#L25) that implements that interface.
```

Comment:
I think it would benefit the reader if we're more precise here, we have an interface `fleet.Service` and an struct `service.Service` that implements the interface.